apps/emqx_plugin_libs/src/emqx_slow_subs/emqx_slow_subs.erl

```diff
@@ -0,0 +1,304 @@
+                              }.
+
+-type stats_update_env() :: #{max_size := pos_integer()}.
+
+-ifdef(TEST).
+-define(TOPK_ACCESS, public).
+-define(EXPIRE_CHECK_INTERVAL, timer:seconds(1)).
+-else.
+-define(TOPK_ACCESS, protected).
+-define(EXPIRE_CHECK_INTERVAL, timer:seconds(10)).
+-endif.
+
+-define(NOW, erlang:system_time(millisecond)).
+-define(NOTICE_TOPIC_NAME, "slow_subs").
+
+%% erlang term order
+%% number < atom < reference < fun < port < pid < tuple < list < bit string
+
+%% ets ordered_set is ascending by term order
+
+%%--------------------------------------------------------------------
+%%--------------------------------------------------------------------
+%% APIs
+%%--------------------------------------------------------------------
+%% @doc Start the st_statistics
+-spec(start_link(Env :: list()) -> emqx_types:startlink_ret()).
+start_link(Env) ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [Env], []).
+
+-spec on_stats_update(stats_update_args(), stats_update_env()) -> true.
+on_stats_update(#{clientid := ClientId,
+                  elapsed := Elapsed,
+                  type := Type,
+                  last_insert_value := LIV,
+                  timestamp := Ts},
+                #{max_size := MaxSize}) ->
+
+    Index = ?INDEX(Elapsed, ClientId),
+
+    case ets:info(?TOPK_TAB, size) of
+        Size when Size < MaxSize - 1, LIV =:= 0 ->
+            %% if the size is enough and it is the first time this client try to insert, insert it directly
+            ets:insert(?TOPK_TAB,
+                       #top_k{index = Index, type = Type, timestamp = Ts});
+        _Size ->
+            case ets:first(?TOPK_TAB) of
+                '$end_of_table' ->
+                    %% if there are no elements, insert directly
+                    ets:insert(?TOPK_TAB,
+                               #top_k{index = Index, type = Type, timestamp = Ts});
+                First ->
+                    %% this client may already be in topk, try to delete
+                    try_delete(First, LIV, ClientId),
+
+                    case Elapsed =< First of
+                        true -> true;
+                        _ ->
+                            %% update record, and remove the first on
+                            ets:insert(?TOPK_TAB,
+                                       #top_k{index = Index, type = Type, timestamp = Ts}),
+
+                            ets:delete(ets:first(?TOPK_TAB))
+                    end
+            end
+    end.
+
+clear_history() ->
+    gen_server:call(?MODULE, ?FUNCTION_NAME).
+
+enable() ->
+    gen_server:call(?MODULE, {enable, true}).
+
+disable() ->
+    gen_server:call(?MODULE, {enable, false}).
+
+%%--------------------------------------------------------------------
+%% gen_server callbacks
+%%--------------------------------------------------------------------
+
+init([Conf]) ->
+    erlang:process_flag(trap_exit, true),
+    init_topk_tab(Conf),
```

Comment:
is it ok for this process to own the table ? if this process restarts, all the inserts will crash with `badarg`